### PR TITLE
[IMPROVEMENT] Use Rest API for file upload

### PR DESCRIPTION
**Pull request by pranavpandey1998official**
removed rn-fetch-blob and use native XMLHttpRequest instead

### CODE REVIEW

**Comment by diegolmello** [comment]
You don't want to append `tmid` every time.
See what happened uploading to a normal conversation.
```
[
  "err_send_image",
  {
    "success": false,
    "error": "Cannot read property 'tmid' of undefined"
  }
]
```

File: app/lib/methods/sendFileMessage.js [hunk 1]
```diff
@@ -55,57 +48,77 @@ export async function sendFileMessage(rid, fileInfo, tmid) {
...
+				fileInfo.progress = Math.floor((loaded / total) * 100);
+				try {
+					database.create('uploads', fileInfo, true);
+				} catch (e) {
+					return log('err_send_file_message_create_upload_2', e);
+				}
+			});
+		};
+
+		xhr.addEventListener('progress', handleProgress);
```
Comment (diegolmello) [inline app/lib/methods/sendFileMessage.js]: This doesn't work.
You should use `xhr.upload.onprogress`.
Example:
```jsx
xhr.upload.onprogress = (event) => {
	console.log('progress: ' + event.loaded + ' / ' + event.total);
}
```

Reference: https://github.com/facebook/react-native/pull/17312

File: app/lib/methods/sendFileMessage.js [hunk 2]
```diff
@@ -55,57 +48,77 @@ export async function sendFileMessage(rid, fileInfo, tmid) {
...
+					database.create('uploads', fileInfo, true);
+					reject(e);
+				} catch (err) {
+					reject(err);
+					log('err_send_file_message_create_upload_3', err);
+				}
+			});
+		};
+
+		xhr.send(formData);
```
Comment (diegolmello) [inline app/lib/methods/sendFileMessage.js]: It's not working for audio messages.

Comment (pranavpandey1998official) [reply app/lib/methods/sendFileMessage.js]: fixed

File: app/lib/methods/sendFileMessage.js (hunk 1 above)
Comment (pranavpandey1998official) [reply app/lib/methods/sendFileMessage.js]: done

File: app/lib/methods/sendFileMessage.js [hunk 3]
```diff
@@ -1,50 +1,43 @@
...
-		const data = await RNFetchBlob.wrap(fileInfo.path);
-		if (!fileInfo.size) {
-			const fileStat = await RNFetchBlob.fs.stat(fileInfo.path);
-			fileInfo.size = fileStat.size;
-			fileInfo.name = fileStat.filename;
-		}
-
-		const { FileUpload_MaxFileSize } = reduxStore.getState().settings;
+export function sendFileMessage(rid, fileInfo, tmid) {
+	return new Promise((resolve, reject) => {
```
[the remark this answers is not in the document]
Comment (pranavpandey1998official) [reply app/lib/methods/sendFileMessage.js]: done

File: app/lib/methods/sendFileMessage.js [hunk 4]
```diff
@@ -1,111 +1,129 @@
-import RNFetchBlob from 'rn-fetch-blob';
-
 import reduxStore from '../createStore';
 import database from '../realm';
 import log from '../../utils/log';
 
-const promises = {};
+const XHR = {};
```
Comment (diegolmello) [inline app/lib/methods/sendFileMessage.js]: Rename it to `uploadQueue`

File: app/containers/MessageBox/Recording.js [hunk 5]
```diff
@@ -44,13 +44,15 @@ export default class extends React.PureComponent {
...
+		this.name = '';
 		this.state = {
 			currentTime: '00:00'
 		};
 	}
 
 	componentDidMount() {
-		const audioPath = `${ AudioUtils.CachesDirectoryPath }/${ Date.now() }.aac`;
+		this.name = `${ Date.now() }.aac`;
+		const audioPath = `${ AudioUtils.CachesDirectoryPath }/${ this.name }`;
```
Comment (diegolmello) [inline app/containers/MessageBox/Recording.js]: It didn't work on Android.
```
[
  "err_finish_audio_message",
  {
    "type": "error",
    "target": {
      "UNSENT": 0,
      "OPENED": 1,
      "HEADERS_RECEIVED": 2,
      "LOADING": 3,
      "DONE": 4,
      "readyState": 4,
      "status": 0,
      "timeout": 0,
      "withCredentials": true,
      "upload": {},
      "_aborted": false,
      "_hasError": true,
      "_method": "POST",
      "_response": "Could not retrieve file for uri /data/user/0/chat.rocket.reactnative/cache/1561671192611.aac",
      "_url": "https://open.rocket.chat/api/v1/rooms.upload/bMvbehmLppt3BzeMcy8bd77ptZswPj3EW8",
      "_timedOut": false,
      "_trackingName": 1022,
      "_incrementalEvents": true,
      "_requestId": null,
      "_headers": {
        "x-auth-token": "eUmO0NS_0iQMI74xzE_wbrlYyTYjhsREbNYsS8pElqC",
        "x-user-id": "y8bd77ptZswPj3EW8"
      },
      "_responseType": "",
      "_sent": true,
      "_lowerCaseResponseHeaders": {},
      "_subscriptions": []
    },
    "currentTarget": {
      "UNSENT": 0,
      "OPENED": 1,
      "HEADERS_RECEIVED": 2,
      "LOADING": 3,
      "DONE": 4,
      "readyState": 4,
      "status": 0,
      "timeout": 0,
      "withCredentials": true,
      "upload": {},
      "_aborted": false,
      "_hasError": true,
      "_method": "POST",
      "_response": "Could not retrieve file for uri /data/user/0/chat.rocket.reactnative/cache/1561671192611.aac",
      "_url": "https://open.rocket.chat/api/v1/rooms.upload/bMvbehmLppt3BzeMcy8bd77ptZswPj3EW8",
      "_timedOut": false,
      "_trackingName": 1022,
      "_incrementalEvents": true,
      "_requestId": null,
      "_headers": {
        "x-auth-token": "eUmO0NS_0iQMI74xzE_wbrlYyTYjhsREbNYsS8pElqC",
        "x-user-id": "y8bd77ptZswPj3EW8"
      },
      "_responseType": "",
      "_sent": true,
      "_lowerCaseResponseHeaders": {},
      "_subscriptions": []
    },
    "eventPhase": 2,
    "bubbles": false,
    "cancelable": false,
    "timeStamp": 1561671198247,
    "isTrusted": false
  }
]
```

File: app/containers/MessageBox/Recording.js [hunk 6]
```diff
@@ -44,13 +44,15 @@ export default class extends React.PureComponent {
 
 		this.recordingCanceled = false;
 		this.recording = true;
+		this.name = '';
```
Comment (diegolmello) [inline app/containers/MessageBox/Recording.js]: Move line 54 to this one.
```
this.name = `${ Date.now() }.aac`;
```

**Pull request merged**